File: media_pipe/sports_talent_app_setup/lib/exercise_engine.dart
import 'package:google_mlkit_pose_detection/google_mlkit_pose_detection.dart';

enum ExerciseType { verticalJump, pushUp, squat }

class ExerciseStatus {
  const ExerciseStatus({
    required this.primaryText,
    this.secondaryText,
    required this.isBodyVisible,
  });

  final String primaryText;
  final String? secondaryText;
  final bool isBodyVisible;
}

abstract class ExerciseEngine {
  ExerciseStatus get status;

  void reset();

  void processPose(Pose? pose);

  /// Raw image-space Y of an adaptive on-screen reference guide, or null if
  /// the engine doesn't use one. Engines that don't need one inherit this
  /// harmless default.
  double? get groundGuideImageY => null;

  /// Whether the UI should show a static "stand here" position guide over
  /// the camera preview right now. Engines that don't need one inherit this
  /// harmless default.
  bool get showPositionGuide => false;

  /// TEMPORARY: multi-line diagnostic text to surface on screen, or null if
  /// the engine has none to show. Engines that don't need this inherit the
  /// harmless default. Remove once no longer needed for physical-test
  /// debugging.
  String? get diagnosticText => null;
}

File: react_folder/src/mediapipe/engines/exerciseEngine.js
/**
 * Base exercise engine — ported from exercise_engine.dart.
 *
 * Defines the exercise type enum, status object, and abstract engine
 * interface that all specific engines implement.
 */

export const ExerciseType = Object.freeze({
  VERTICAL_JUMP: 'verticalJump',
  PUSH_UP: 'pushUp',
  SQUAT: 'squat',
  CONTROLLED_CRUNCH: 'controlledCrunch',
  WALL_SIT: 'wallSit',
  PLANK: 'plank',
});

export class ExerciseStatus {
  constructor({ primaryText, secondaryText = null, isBodyVisible }) {
    this.primaryText = primaryText;
    this.secondaryText = secondaryText;
    this.isBodyVisible = isBodyVisible;
  }
}

/**
 * Abstract base for all exercise engines.
 * Each engine receives per-frame pose data and maintains its own state machine.
 */
export class ExerciseEngine {
  get status() {
    return new ExerciseStatus({ primaryText: '', isBodyVisible: false });
  }

  reset() {}

  /** @param {Array|null} landmarks  MediaPipe normalized landmarks or null */
  processLandmarks(_landmarks) {}

  /** Raw normalized Y of an adaptive ground guide, or null. */
  get groundGuideY() { return null; }

  /** Whether the UI should show a static position guide. */
  get showPositionGuide() { return false; }
}
